File: apps/cinder/src/handlers/heartbeat.ts
import { pluginInstallations, projects, serverMetadata } from '@bx-team/stratus/d1';
import type { Heartbeat } from '@bx-team/types/schema/pulsify';
import { and, inArray, sql } from 'drizzle-orm';
import type { HandlerContext } from '../env';
import { writeServerStats } from '../lib/analytics';

export async function handleHeartbeat(ctx: HandlerContext, event: Heartbeat, projectId: string, countryCode: string) {
  const { db, env } = ctx;
  const now = new Date();

  await db
    .insert(serverMetadata)
    .values({
      projectId,
      software: event.server.software,
      mcVersion: event.server.version,
      countryCode,
      lastSeenAt: now,
    })
    .onConflictDoUpdate({
      target: serverMetadata.projectId,
      set: {
        software: event.server.software,
        mcVersion: event.server.version,
        countryCode,
        lastSeenAt: now,
      },
    });

  writeServerStats(env, projectId, {
    online: event.server.online,
    tps: event.server.tps,
    mspt: event.server.mspt,
    memoryUsed: event.server.memory_used_mb,
    memoryMax: event.server.memory_max_mb,
  });

  try {
    if (event.plugins.length > 0) {
      const infoByName = new Map(event.plugins.map(p => [p.name, p]));
      const matchedPlugins = await db
        .select({ id: projects.id, name: projects.name })
        .from(projects)
        .where(and(inArray(projects.name, [...infoByName.keys()]), inArray(projects.type, ['plugin', 'mod'])));

      const rows = matchedPlugins.flatMap(pp => {
        const info = infoByName.get(pp.name);
        if (!info) return [];
        return [
          {
            pluginId: pp.id,
            serverId: projectId,
            version: info.version,
            enabled: info.enabled,
            shareErrors: true,
            lastSeenAt: now,
          },
        ];
      });

      if (rows.length > 0) {
        await db
          .insert(pluginInstallations)
          .values(rows)
          .onConflictDoUpdate({
            target: [pluginInstallations.pluginId, pluginInstallations.serverId],
            set: {
              version: sql`excluded.version`,
              enabled: sql`excluded.enabled`,
              lastSeenAt: now,
            },
          });
      }
    }
  } catch (err) {
    console.error('[heartbeat] Failed to update plugin installations:', err);
  }
}

File: apps/cinder/src/handlers/heartbeat.rs
use analytics::writer::ServerStatsPoint;
use database::Transaction;
use database::models::pulsify;
use database::queue::QueuedEvent;
use sqlx::Acquire;
use types::ingest::Heartbeat;

use super::Outcome;
use crate::Error;
use crate::state::AppState;

pub async fn handle(
    state: &AppState,
    tx: &mut Transaction<'_>,
    message: &QueuedEvent,
    heartbeat: &Heartbeat,
    outcome: &mut Outcome,
) -> Result<(), Error> {
    let country = state.country_of(message.ip.as_deref());

    pulsify::upsert_server_metadata(
        tx,
        message.project_id,
        &heartbeat.server.software,
        &heartbeat.server.version,
        &country,
    )
    .await?;

    outcome.analytics.server_stats.push(ServerStatsPoint {
        timestamp: message.received_at,
        project_id: message.project_id,
        online: u32::try_from(heartbeat.server.online.max(0)).unwrap_or(u32::MAX),
        tps: heartbeat.server.tps,
        mspt: heartbeat.server.mspt,
        memory_used_mb: u64::try_from(heartbeat.server.memory_used_mb.max(0)).unwrap_or(0),
        memory_max_mb: u64::try_from(heartbeat.server.memory_max_mb.max(0)).unwrap_or(0),
    });

    record_installations(tx, message, heartbeat).await;
    Ok(())
}

/// Installation matching runs in its own savepoint: a plugin name clash must not cost the
/// heartbeat that carried it.
async fn record_installations(
    tx: &mut Transaction<'_>,
    message: &QueuedEvent,
    heartbeat: &Heartbeat,
) {
    if heartbeat.plugins.is_empty() {
        return;
    }

    let mut names = Vec::with_capacity(heartbeat.plugins.len());
    let mut versions = Vec::with_capacity(heartbeat.plugins.len());
    let mut enabled = Vec::with_capacity(heartbeat.plugins.len());
    for plugin in &heartbeat.plugins {
        names.push(plugin.name.clone());
        versions.push(plugin.version.clone());
        enabled.push(plugin.enabled);
    }

    let savepoint = match tx.begin().await {
        Ok(savepoint) => savepoint,
        Err(error) => {
            tracing::warn!(%error, "could not open a savepoint for installations");
            return;
        }
    };
    let mut savepoint = savepoint;

    match pulsify::upsert_installations(
        &mut savepoint,
        message.project_id,
        &names,
        &versions,
        &enabled,
    )
    .await
    {
        Ok(_) => {
            if let Err(error) = savepoint.commit().await {
                tracing::warn!(%error, "could not record plugin installations");
            }
        }
        Err(error) => {
            tracing::warn!(%error, project_id = %message.project_id, "plugin matching failed");
            let _ = savepoint.rollback().await;
        }
    }
}
